testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketSslClientRenegotiateTest.java

```diff
@@ -0,0 +1,223 @@
+
+import static org.junit.Assert.*;
+
+@RunWith(Parameterized.class)
+public class SocketSslClientRenegotiateTest extends AbstractSocketTest {
+
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(
+            SocketSslClientRenegotiateTest.class);
+    private static final File CERT_FILE;
+    private static final File KEY_FILE;
+
+    static {
+        SelfSignedCertificate ssc;
+        try {
+            ssc = new SelfSignedCertificate();
+        } catch (CertificateException e) {
+            throw new Error(e);
+        }
+        CERT_FILE = ssc.certificate();
+        KEY_FILE = ssc.privateKey();
+        System.setProperty("jdk.tls.rejectClientInitiatedRenegotiation", "true");
+    }
+
+    @Parameters(name =
+            "{index}: serverEngine = {0}, clientEngine = {1}")
+    public static Collection<Object[]> data() throws Exception {
+        List<SslContext> serverContexts = new ArrayList<SslContext>();
+        if (PlatformDependent.javaVersion() >= 8) {
+            // Only supported on java8+.
+            serverContexts.add(new JdkSslServerContext(CERT_FILE, KEY_FILE));
+        }
+
+        List<SslContext> clientContexts = new ArrayList<SslContext>();
+        clientContexts.add(new JdkSslClientContext(CERT_FILE));
+
+        boolean hasOpenSsl = OpenSsl.isAvailable();
+        if (hasOpenSsl) {
+            OpenSslServerContext context = new OpenSslServerContext(CERT_FILE, KEY_FILE);
+            serverContexts.add(context);
+        } else {
+            logger.warn("OpenSSL is unavailable and thus will not be tested.", OpenSsl.unavailabilityCause());
+        }
+
+        List<Object[]> params = new ArrayList<Object[]>();
+        for (SslContext sc: serverContexts) {
+            for (SslContext cc: clientContexts) {
+                    for (int i = 0; i < 32; i++) {
+                        params.add(new Object[] {
+                                sc, cc});
+                    }
+            }
+        }
+
+        return params;
+    }
+
+    private final SslContext serverCtx;
+    private final SslContext clientCtx;
+
+    private final AtomicReference<Throwable> clientException = new AtomicReference<Throwable>();
+    private final AtomicReference<Throwable> serverException = new AtomicReference<Throwable>();
+
+    private volatile Channel clientChannel;
+    private volatile Channel serverChannel;
+
+    private volatile SslHandler clientSslHandler;
+    private volatile SslHandler serverSslHandler;
+
+    private final TestHandler clientHandler = new TestHandler(clientException);
+
+    private final TestHandler serverHandler = new TestHandler(serverException);
+
+    public SocketSslClientRenegotiateTest(
+            SslContext serverCtx, SslContext clientCtx) {
+        this.serverCtx = serverCtx;
+        this.clientCtx = clientCtx;
+    }
+
+    @AfterClass
+    public static void after() {
+        System.setProperty("jdk.tls.rejectClientInitiatedRenegotiation", "false");
```

Comment:
Is this necessary?  And if we are trying to "set it back" should we save what the value was before we set it?